rfcs/0006-transaction.md

```diff
@@ -0,0 +1,366 @@
+    write_keys: HashSet<Bytes>,
+    read_keys: Vec<Range<Bytes>>,
+    committed_seq: Option<u64>
+}
+```
+
+`Oracle` is considered a global singleton in the DB, and it tracks the recent committed transactions in the `recent_committed_txns` deque. When a transaction commits, it should push itself into the `recent_committed_txns` deque, and the entries in `recent_committed_txns` can be GCed after they are no longer needed. Also, when a `TransactionState` is tracked in the `recent_committed_txns`, only the write keys need to be tracked, while `read_keys` should always be empty.
+
+Besides the started sequence number of the transaction, we also need to track the sequence number of the transaction committed. During the execution of a transaction, some other transactions might commit, and the current transaction should check the conflicts with these committed transactions.
+
+The conflict check simply checks whether the Read & Write keys of the current transaction are overlapped with the Write keys from related recent committed transactions.
+
+Given an execution history of transactions like:
+
+```
+txn1: [seq:97] <--- ---- ---> [seq:99]
+txn2:     [seq:98]  <--- ---- ---- ---- ---- ---- ---> [seq:105]
+txn3:                         [seq:102] <--- ---> [seq:104]
+txn4:                    [seq:101] <--- ---- ---- ---- ---> [seq:106]
+```
+
+Let's say the current transaction is `txn4`, it should check conflicts with `txn2` and `txn3` during the commit, because `txn2` and `txn3` are committed after `txn4` started. During the conflict checking, `txn4`'s Read & Write keys should not conflict with the Write keys of `txn2` and `txn3`.
+
+It does not need to check conflicts with `txn1`, because `txn1` is committed before `txn4` started. This could tell us when to GC the entry in the `recent_committed_txns` deque: when ALL the running transactions' sequence number are bigger than the committed sequence number of the entry, then the entry could be GCed.
+
+Please note that `read_keys` are represented as a set of `Range<Bytes>`. This allows us to efficiently track ranges of keys that were read during the transaction, and ensure Serializable Snapshot Isolation (SSI) on range conflicts. For example, if transaction A reads a range `["key01", "key10"]` and transaction B writes to `"key05"` which falls within that range, we can detect this conflict and abort one of the transactions to maintain serializability.
+
+This requires we track the original write keys in the `TransactionState`, instead of the key fingerprints like Badger does. We could reuse the buffers of the original keys in WriteBatch to reference the write keys with slices of `Bytes`.
+
+The non-transactioned writes operations should also be tracked in the `TransactionState` and `recent_committed_txns` for conflict checking. They could be regarded as a transaction that only contains a single write operation, and the start sequence number of this transaction is same as the committed sequence number.
```

Comment:
Would there be a different `open` option, that returns a DB from which a transaction can be created?
Tracking non-transactional writes as well would mean additional state to be tracked, and it is required only when transactions are involved. Having an explicit option would mean that this tracking is enabled only when there is an explicit intent from user.

Reply:
Perhaps we could track only when there is at least one open transaction. And then we don't need a different kind of a DB.

Reply:
i hope this will not become a bottleneck, and it needs prove not a fallback in the benchmark.

there's a double linked list to track the active Snapshot in leveldb, while each Snapshot is created whenever any `range()` started: https://github.com/google/leveldb/blob/main/db/snapshot.h#L30

having double linked list is the toughest part from the language, but we can find some alternative way to track this list in the `Oracle`. the only purpose to track the read snapshots are to get the smallest sequence number, which to tell the compactors know which entries are safe to be reclaimed.

Reply:
Yeah I'm inclined to see if it's actually a problem first. If it is, then I like @vigneshc's idea about only tracking writes when there are actually open transactions that the write could conflict with.